runtime/expr/function/unflatten.go

```diff
@@ -140,3 +88,67 @@ func (u *Unflatten) decodeKey(b zcode.Bytes) field.Path {
 	}
 	return u.path
 }
+
+type recordCache struct {
+	index   int
+	records []*record
+}
+
+func (c *recordCache) new() *record {
+	if c.index == cap(c.records) {
+		c.records = append(c.records, new(record))
+	}
+	r := c.records[c.index]
+	r.columns = r.columns[:0]
+	r.records = r.records[:0]
+	c.index++
+	return r
+}
+
+func (c *recordCache) reset() {
+	c.index = 0
+}
+
+type record struct {
+	columns []zed.Column
+	records []*record
+}
+
+func (r *record) addPath(c *recordCache, p []string) {
+	if len(p) == 0 {
+		return
+	}
+	if len(r.columns) == 0 || r.columns[len(r.columns)-1].Name != p[0] {
+		r.appendColumn(p[0])
```

Comment:
Isn't this adding an allocation whereas appendColumns amortizes that cost over time?

Reply:
Not here. The compiler is generally smart enough to avoid allocating a new backing array for a slice on every `append`, and we're appending a `zed.Column` here, so that doesn't require a separate allocation in the way that appending a `*zed.Column` would.

The program below shows `append`'s usual behavior, which is pretty consistent regardless of `slice`'s type. (The exception I've seen is `[]struct{}`, for which capacity always increases by one after `append` with Go 1.18.1.)
```go
package main

import (
	"fmt"
)

func main() {
	var slice []int
	lastcap := cap(slice)
	for {
		slice = append(slice, 0)
		if currentcap := cap(slice); currentcap != lastcap {
			fmt.Println("cap(slice) changed to", currentcap)
			lastcap = currentcap
		}
	}
}
```

Reply:
Makes sense.